We should be caching connectorClients

tomlm opened the pull request.

drub0y commented:
So this certainly works, couple thoughts...

First, at bare minimum I suggest considering switching to use `GetOrAdd` rather than the separate `TryGet`, do some work, `TryAdd`. I personally find it reads clearer, but, more importantly, it will maintain the lock if the "get" fails which you do the work to sping up the client, which helps prevent two threads at once both doing the work and one of them just "failing" at the `TryAdd` and leaving a dangling connection around for some time.
 
That said, I would honestly recommend throwing a proper level of indirection in there in the form of an `IConnectorClientFactory` with a `ValueTask<IConnectorClient> GetClientAsync(string serviceUrl, MicrosoftAppCredentials appCredentials)` method. This will buy you at least a couple of things:
    1. Take the responsibility out of the `BotFrameworkAdapter` with respect to how clients are even _created_, let alone how they might be cached.
    2. Improve testability as this isn't really verifiable in its current form. With the factory approach you'll be able to verify the `BotFrameworkAdapter` calls to gets client instances when it's supposed to and that's it. Then you can also independently test the implementation of any `IConnectorClientFactory` implementations separately.
    3. Allows for other implementations around creating/caching `IConnectorClient`'s to be implemented. You never know what other patterns people might want/need. Also, by returning the interface instead of the `ConnectorClient` class you would give some really good flexibility around the underlying implementation of the client itself in the long run.

Funny side note here: you already have some of this design in the `Classic` project. I wouldn't necessarily use that API design directly, but I definitely think it was definitely factored out more cleanly there.

tomlm commented:
I have changed to use GetOrUpdate() 
I have also changed the other cahce in this class to use ConcurrentDictionary<>
I have filed a DCR #883 to discuss whether we should move to a ClassFactory model for connector Client
